tmf/org.eclipse.tracecompass.tmf.ui/src/org/eclipse/tracecompass/tmf/ui/widgets/timegraph/widgets/TimeGraphControl.java

```diff
@@ -2333,8 +2334,16 @@ protected void drawMarker(IMarkerEvent marker, Rectangle bounds, ITimeDataProvid
             case SymbolType.CIRCLE:
                 SymbolHelper.drawCircle(gc, color, symbolSize, rect.x, y);
                 break;
-            default:
+            case SymbolType.DIAMOND:
                 SymbolHelper.drawDiamond(gc, color, symbolSize, rect.x, y);
+                break;
+            default:
+                Color oldColor = gc.getForeground();
+                gc.setForeground(color);
+                int height = (int) (rect.height * heightFactor);
+                TimeGraphRender.setFontForHeight(height, gc);
+                gc.drawText(symbol, rect.x - symbolSize, y - symbolSize, true);
```

Comment:
For some reason, the height of the text bounds is not actually what was requested [*], so its half-height is not equal to symbolSize. Unfortunately we must compute the text height and use it not only to adjust the top of text bounds relative to the requested middle (y) but we must also recalculate the middle y for TOP and BOTTOM alignment.

This works for me:

                int textSize = (gc.textExtent(symbol).y + 1) / 2;
                if (VerticalAlign.BOTTOM.equals(verticalAlign)) {
                    y = rect.y + rect.height / 2 + Math.max(0, rect.height / 2 - textSize);
                } else if (VerticalAlign.TOP.equals(verticalAlign)) {
                    y = rect.y + rect.height / 2 - Math.max(0, rect.height / 2 - textSize);
                }
                gc.drawText(symbol, rect.x - symbolSize, y - textSize, true);

[*] In my debugging, when we request a font with height of 100 pixels, this corresponds to a requested font data height of 75 points (with DPI=96 px/in and PPI=72 pts/in), but the created font data's height is 113 (apparently in pixels although the API says this should be in points), and the displayed text indeed is 113 pixels in height.

Reply:
Oh yeah, it is much more apparent when zooming in.

Does this only replace `gc.drawText()`, or does it also replace `TimeGraphRender.setFontForHeight()`? I don't think `TimeGraphRender.setFontForHeight()` is actually doing anything.

Reply:
The code should replace only line 2345. TimeGraphRender.setFontForHeight() will set the font on the gc for the following drawText operation(s). It will either create (first time) or reuse a font with the expected height.

Reply:
Sounds good, that's what I did :grin: